packages/blocks/src/api/raw-handling/html-formatting-remover.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * Internal dependencies
+ */
+import { isPhrasingContent } from './phrasing-content';
+
+function getSibling( node, which ) {
+	const sibling = node[ `${ which }Sibling` ];
+
+	if ( sibling && isPhrasingContent( sibling ) ) {
+		return sibling;
+	}
+
+	const { parentNode } = node;
+
+	if ( ! parentNode || ! isPhrasingContent( parentNode ) ) {
+		return;
+	}
+
+	return getSibling( parentNode, which );
+}
+
+function isFormattingSpace( character ) {
+	return (
+		character === ' ' ||
+		character === '\r' ||
+		character === '\n' ||
+		character === '\t'
+	);
+}
+
+/**
+ * Removes spacing that formats HTML.
+ *
+ * @see https://www.w3.org/TR/css-text-3/#white-space-processing
+ *
+ * @param {Node} node The node to be processed.
+ * @return {void}
+ */
+export default function( node ) {
+	if ( node.nodeType !== node.TEXT_NODE ) {
+		return;
+	}
+
+	// Ignore pre content.
+	if ( node.parentElement.closest( 'pre' ) ) {
```

Comment:
Is there a chance `node.parentElement` is `null`?

Reply:
I guess it works because the HTML is inserted into a `<body>` tag, so there will always at least be `body` as the `parentElement`.

https://github.com/WordPress/gutenberg/blob/ada00b67d53d505aee41e774e0f4b72973de9d5a/packages/blocks/src/api/raw-handling/utils.js#L161

Aside: An interesting consequence of this is that trying to filter HTML which includes a `<body>` tag doesn't work quite as expected. Not sure it's something that would ever happen in a real-world scenario though.

Reply:
It's always wrapped in at least a `body` element.

Reply:
Since it's an implementation detail, it could help to have a test case which verifies the behavior of filtering on a plain string of text.

Reply:
Added an empty string test.

Reply:
Hm, I guess that makes not difference. Adjusting :)